### PR TITLE
[MOB-18293] - Handling Notification Actions

### PR DESCRIPTION
**Pull request by PravinPK**
### Definition
Notification Action is defined as an action that the Messaging Extension takes in response to interacting with the notification.
Typically these actions can be to open a web URL or a deeplink associated with that application.

### How notification action is set
While designing a push notification campaign. Notification action is set by editing the click behavior field.
In iOS deeplink/ weburl action links are sent to the device through the push notification payload. They are available in a custom field "adb_uri".

Screenshot 2023-04-11 at 11 52 56 AM. 

### Notification body click action 
Notification body click action determines the action taken when clicking on the body of the notification. This will typically open the application and hence requires the customers to call  Messaging API handleNotificationResponse with  applicationOpened : true
customActionId : nil
`Messaging.handleNotificationResponse(response, applicationOpened: true, customActionId: nil)`

###  Notification custom button click action
Notification custom button click action determines the action taken when the user clicks either of the action buttons in a notification. On this action we require the customers to call Messaging API handleNotificationResponse with
applicationOpened : true/false (depending upon if the action was set to bring the application to foreground)
customActionId : response.actionIdentifier

`Messaging.handleNotificationResponse(response, applicationOpened: false/true, customActionId: response.actionIdentifier)`

image

### Why cant we handle notification custom button click action
With the [examples](https://github.com/adobe/aepsdk-messaging-ios/compare/main...PravinPK:aepsdk-messaging-ios:dev-v1.1.1?expand=1#diff-7b5f15992292d4c8a9c5241e39c2f11afced0fc54a089fb8d0b3492105e08c4eR143-R180) in this PR. The notification custom actions along with its category is pre-created by the application and registered at launch time. Therefore AJO doesn't control the creation of action buttons from its authoring UI. The authoring UI should only ask for the categoryID of the already registered category that can be invoked. 

Since AJO doesn't have the power to define the actions of a custom notification button. So Messaging SDK would not be able to handle the custom notification action links. This has to be handled by the application as shown [here](https://github.com/adobe/aepsdk-messaging-ios/compare/main...PravinPK:aepsdk-messaging-ios:dev-v1.1.1?expand=1#diff-7b5f15992292d4c8a9c5241e39c2f11afced0fc54a089fb8d0b3492105e08c4eR184-R196).

### A Complex Scenario
Assume an AJO consumer wants to design the following notification such that

Screenshot 2023-04-11 at 1 52 07 PM

1. Clicking on the body should open a deeplink to the tutorial
2. Clicking on the custom action "Read Later" should dismiss the notification
3. Clicking on the custom action  "Show Details" should open the app with a deeplink
4. Clicking on the custom action "Unsubscribe" should open the app with deeplink to unsubscribe

To achieve this example:
1. The app should be predefined to register a notification category called "tutorial" with 3 buttons with names as shown above
2. The app should also be predefined to handle the action when each of these action buttons is clicked

3. Now the user goes to AJO UI and sets the notification content with
        Body click behavior - DEEPLINK  (with deeplink of that specific tutorial) 
        iOS categoryID - "tutorial" (this will show the 3 button that are preconfigured)
       + any custom notification payload that are necessary

4.  In this example the messaging SDK should only handle the action of the notification body but not the 3 action buttons. Hence in this PR the notification interaction event includes the "pushclickthroughurl" in eventData only when the API is passed with customaction `nil` and `applicationOpened` = true ([code here](https://github.com/adobe/aepsdk-messaging-ios/pull/174/files#))

### CODE REVIEW

File: AEPMessaging/Sources/Messaging+PublicAPI.swift [hunk 1]
```diff
@@ -42,6 +42,17 @@ import UserNotifications
                                         MessagingConstants.XDM.Key.ADOBE_XDM: xdm ?? [:]] // If xdm data is nil we use empty dictionary
         if customActionId == nil {
             eventData[MessagingConstants.Event.Data.Key.EVENT_TYPE] = MessagingConstants.XDM.Push.EventType.APPLICATION_OPENED
+            
+            // Messaging extension can handle click through url's when the user taps on the notification
+            // Messaging extension CANNOT handle click through url's when the user taps on any of the action buttons by swiping the notifications. They must be handled in the application.
+            if (applicationOpened) {
+                // Add the click thorough url to the eventData if the application is opened and when the customActionId is nil                
+                // (TODO: Decide if we should also accept UNNotificationDefaultActionIdentifier as customActionId to process sending the clickthrough url through eventData)
+                if let actionUrl = notificationRequest.content.userInfo [MessagingConstants.PushNotification.UserInfoKey.ACTION_URL] as? String {
```
Comment (sbenedicadb) [inline AEPMessaging/Sources/Messaging+PublicAPI.swift:51]: nit: no space between dictionary and subscript `userInfo[Messaging...`

File: AEPMessaging/Sources/Messaging+PublicAPI.swift [hunk 2]
```diff
@@ -42,6 +42,17 @@ import UserNotifications
                                         MessagingConstants.XDM.Key.ADOBE_XDM: xdm ?? [:]] // If xdm data is nil we use empty dictionary
         if customActionId == nil {
             eventData[MessagingConstants.Event.Data.Key.EVENT_TYPE] = MessagingConstants.XDM.Push.EventType.APPLICATION_OPENED
+            
+            // Messaging extension can handle click through url's when the user taps on the notification
+            // Messaging extension CANNOT handle click through url's when the user taps on any of the action buttons by swiping the notifications. They must be handled in the application.
+            if (applicationOpened) {
+                // Add the click thorough url to the eventData if the application is opened and when the customActionId is nil                
+                // (TODO: Decide if we should also accept UNNotificationDefaultActionIdentifier as customActionId to process sending the clickthrough url through eventData)
```
Comment (sbenedicadb) [inline AEPMessaging/Sources/Messaging+PublicAPI.swift:50]: do we need to chat about this?

File: AEPMessaging/Sources/Messaging.swift [hunk 3]
```diff
@@ -277,9 +278,19 @@ public class Messaging: NSObject, Extension {
 
         // Check if the event type is `MessagingConstants.Event.EventType.messaging` and
         // eventSource is `EventSource.requestContent` handle processing of the tracking information
-        if event.isMessagingRequestContentEvent, configSharedState.keys.contains(MessagingConstants.SharedState.Configuration.EXPERIENCE_EVENT_DATASET) {
-            handleTrackingInfo(event: event)
-            return
+        if event.isMessagingRequestContentEvent {
+            
+            //  Handle push click through action if the event data contains the web/deeplink
```
Comment (sbenedicadb) [inline AEPMessaging/Sources/Messaging.swift:283]: click-through

File: AEPMessaging/Sources/Messaging.swift [hunk 4]
```diff
@@ -277,9 +278,19 @@ public class Messaging: NSObject, Extension {
...
+            if let actionLink = event.pushClickThroughLink {
+                DispatchQueue.main.async {
+                    UIApplication.shared.open(actionLink)
+                }                
+            }
+            
+            if(configSharedState.keys.contains(MessagingConstants.SharedState.Configuration.EXPERIENCE_EVENT_DATASET)) {
+                handleTrackingInfo(event: event)
+                return
+            }
```
Comment (sbenedicadb) [inline AEPMessaging/Sources/Messaging.swift:293]: should we swap the order of these? did we test to make sure the tracking call still goes through even after navigating away potentially to another app?

File: AEPMessaging/Sources/Messaging.swift [hunk 5]
```diff
@@ -277,9 +278,19 @@ public class Messaging: NSObject, Extension {
...
+        if event.isMessagingRequestContentEvent {
+            
+            //  Handle push click through action if the event data contains the web/deeplink
+            if let actionLink = event.pushClickThroughLink {
+                DispatchQueue.main.async {
+                    UIApplication.shared.open(actionLink)
+                }                
+            }
+            
+            if(configSharedState.keys.contains(MessagingConstants.SharedState.Configuration.EXPERIENCE_EVENT_DATASET)) {
```
Comment (sbenedicadb) [inline AEPMessaging/Sources/Messaging.swift:290]: nit: don't use params on `if` statements 😄

File: AEPMessaging/Sources/Event+Messaging.swift [hunk 6]
```diff
@@ -296,6 +296,13 @@ extension Event {
     var actionId: String? {
         data?[MessagingConstants.Event.Data.Key.ACTION_ID] as? String
     }
+    
+    var pushClickThroughLink: URL? {
```
Comment (sbenedicadb) [inline AEPMessaging/Sources/Event+Messaging.swift:300]: maybe rename to `pushClickThroughUrl`?

**Review by sbenedicadb — CHANGES_REQUESTED** [review]

**Comment by PravinPK** [comment]
Closing this PR. Will put up another PR with a new API that can handle actions on notifications

**Pull request closed without merging**